packages/interactor/src/specification.ts

```diff
@@ -1,9 +1,130 @@
+   * to make them more specific by limiting the interactor to a section of the
+   * page.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await Fieldset('Owner').find(TextField('Name')).fillIn('Jonas');
+   * await Fieldset('Brand').find(TextField('Name')).fillIn('Volkswagen');
+   * ```
+   * @param interactor the interactor which should be scoped
+   * @returns a scoped copy of the initial interactor
+   * @typeParam T the type of the interactor that we are going to scope
+   */
+  find<T extends Interactor<any, any>>(interactor: T): T;
+
+  /**
+   * Perform a one-off action on the given interactor. Takes a function which
+   * receives an element. This function converges, which means that it is rerun
+   * in a loop until it does not throw an error or times out.
+   *
+   * We recommend using this function for debugging only. You should normally
+   * define an action in an {@link InteractorSpecification}.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await Link('Next').perform((e) => e.click());
+   * ```
+   */
+  perform(fn: (element: E) => void): Interaction<void>;
+
+  /**
+   * Perform a one-off assertion on the given interactor. Takes a function which
+   * receives an element. This function converges, which means that it is rerun
+   * in a loop until it does not throw an error or times out.
+   *
+   * We recommend using this function for debugging only. You should normally
+   * define a filter in an {@link InteractorSpecification}.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await Link('Next').assert((e) => assert(e.tagName === 'A'));
+   * ```
+   */
+  assert(fn: (element: E) => void): Interaction<void>;
+
+  /**
+   * An assertion which checks that an element matching the interactor exists.
+   * Throws an error if the element does not exist.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await Link('Next').exists();
+   * ```
+   */
+  exists(): ReadonlyInteraction<void>;
+
+  /**
+   * An assertion which checks that an element matching the interactor does not
+   * exist. Throws an error if the element exists.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await Link('Next').absent();
+   * ```
+   */
+  absent(): ReadonlyInteraction<void>;
+
+  /**
+   * Checks that there is one element matching the interactor, and that this
+   * element matches the given filters. The available filters are defined by
+   * the {@link InteractorSpecification}.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await Link('Home').has({ href: '/' })
+   * ```
+   */
+  has(filters: F): ReadonlyInteraction<void>;
+
+  /**
+   * Identical to {@link has}, but reads better with some filters.
+   *
+   * ## Example
+   *
+   * ``` typescript
+   * await CheckBox('Accept conditions').is({ checked: true })
+   * ```
+   */
+  is(filters: F): ReadonlyInteraction<void>;
+}
+
+export type ActionFn<E extends Element> = (interactor: Interactor<E, {}>, ...args: any[]) => Promise<unknown>;
```

Comment:
Oh yeah, I keep forgetting about this! Maybe `never` would be an appropriate choice here, I'll try it out and see if it works!

Reply:
Okay, so I tried this with both `never` and `Record<string, any>` and neither of these work, unfortunately. I don't have any other ideas for this, unfortunately.

`never` I think doesn't work because it breaks the mapped types, `Record<string, any>` is too permissive, because due to the type mapping it will actually allow `myInteractors.completelyArbitraryMethod()`, so basically you can call *any* method on the interactor. Do you have any other ideas for types to try?

Reply:
@jnicklas you could try `Record<string, never>` which really is a type of empty object

`export type ActionFn<E extends Element> = (interactor: Interactor<E, Record<string, never>>, ...args: any[]) => Promise<unknown>;`

Reply:
Interesting! This one did work for the constraints in `ActionFn` and `ActionMethods`, but it does *not* work as the default type in InteractorBuilder. I think this is because this type gets composed into an intersection type like this:

``` typescript
type Result = Record<string, never> & { some: string };
```

Which would effectively expand to something like:


``` typescript
type Result = { some: string & never };
```

Which doesn't make sense.

I'm guessing that using this for the constraints only would be better.

Once again, the lack of a proper "empty object" type in TypeScript is pretty annoying.

Reply:
@jnicklas what should the defaults be for `InteractorBuilder`, would they really be an empty object.  Would they not be `never`?